src/waveform/renderers/allshader/waveformrendererslipmode.cpp

```diff
@@ -1,116 +1,134 @@
 #include "waveform/renderers/allshader/waveformrendererslipmode.h"
 
 #include <QDomNode>
+#include <QVector4D>
 #include <memory>
 
 #include "control/controlproxy.h"
+#include "rendergraph/geometry.h"
+#include "rendergraph/material/rgbamaterial.h"
+#include "rendergraph/vertexupdaters/rgbavertexupdater.h"
+#include "util/colorcomponents.h"
 #include "waveform/renderers/waveformwidgetrenderer.h"
 #include "waveform/waveformwidgetfactory.h"
 #include "widget/wskincolor.h"
 
 namespace {
 
 constexpr int kBlinkingPeriodMillis = 1600;
-// Position matrix passed to OpenGL when drawing the shader
-constexpr float positionArray[] = {-1.f, -1.f, 1.f, -1.f, -1.f, 1.f, 1.f, 1.f};
 
 // Used as default outline color in case no value is provided in the theme
-
-#if QT_VERSION >= QT_VERSION_CHECK(6, 0, 0)
-constexpr
-#else
-const
-#endif
-        QColor kDefaultColor = QColor(224, 224, 224);
+constexpr QColor kDefaultColor = QColor(224, 224, 224);
 
 } // anonymous namespace
 
+using namespace rendergraph;
+
 namespace allshader {
 
 WaveformRendererSlipMode::WaveformRendererSlipMode(
         WaveformWidgetRenderer* waveformWidget)
-        : WaveformRenderer(waveformWidget),
+        : ::WaveformRendererAbstract(waveformWidget),
           m_slipBorderTopOutlineSize(10.f),
           m_slipBorderBottomOutlineSize(10.f) {
+    initForRectangles<RGBAMaterial>(0);
+    setUsePreprocess(true);
+}
+
+void WaveformRendererSlipMode::draw(QPainter* painter, QPaintEvent* event) {
+    Q_UNUSED(painter);
+    Q_UNUSED(event);
+    DEBUG_ASSERT(false);
 }
 
 bool WaveformRendererSlipMode::init() {
     m_timer.restart();
 
-    m_pSlipMode = std::make_unique<ControlProxy>(
-            m_waveformRenderer->getGroup(), "slip_enabled");
+    m_pSlipModeControl.reset(new ControlProxy(
+            m_waveformRenderer->getGroup(), "slip_enabled"));
 
     return true;
 }
 
-void WaveformRendererSlipMode::setup(const QDomNode& node, const SkinContext& context) {
-    const QString slipModeOutlineColorName = context.selectString(node, "SlipBorderOutlineColor");
+void WaveformRendererSlipMode::setup(const QDomNode& node, const SkinContext& skinContext) {
+    const QString slipModeOutlineColorName =
+            skinContext.selectString(node, "SlipBorderOutlineColor");
     if (!slipModeOutlineColorName.isNull()) {
         m_color = WSkinColor::getCorrectColor(QColor(slipModeOutlineColorName));
     } else {
         m_color = kDefaultColor;
     }
-    const float slipBorderTopOutlineSize = context.selectFloat(
+    const float slipBorderTopOutlineSize = skinContext.selectFloat(
             node, "SlipBorderTopOutlineSize", m_slipBorderTopOutlineSize);
```

Comment:
Could you update this string to a literal?

Reply:
done